src/buffer/out/textBuffer.cpp

```diff
@@ -2543,246 +2575,267 @@ void TextBuffer::Serialize(const wchar_t* destination) const
 
     // This iterates through each row. The exit condition is at the end
     // of the for() loop so that we can properly handle file flushing.
-    for (til::CoordType currentRow = 0;; currentRow++)
+    for (til::CoordType currentRow = firstRow;; currentRow++)
     {
         const auto& row = GetRowByOffset(currentRow);
 
-        if (const auto lr = row.GetLineRendition(); lr != LineRendition::SingleWidth)
+        const auto moreRowsRemaining = currentRow < lastRow;
+        const auto startX = 0;
+        const auto endX = row.GetReadableColumnCount();
+        const bool addLineBreak = !row.WasWrapForced() || !moreRowsRemaining;
+
+        SerializeRow(row, startX, endX, addLineBreak, buffer, previousTextAttr);
+
+        if (buffer.size() >= writeThreshold || !moreRowsRemaining)
         {
-            static constexpr std::wstring_view mappings[] = {
-                L"\x1b#6", // LineRendition::DoubleWidth
-                L"\x1b#3", // LineRendition::DoubleHeightTop
-                L"\x1b#4", // LineRendition::DoubleHeightBottom
-            };
-            const auto idx = std::clamp(static_cast<int>(lr) - 1, 0, 2);
-            buffer.append(til::at(mappings, idx));
+            const auto fileSize = gsl::narrow<DWORD>(buffer.size() * sizeof(wchar_t));
+            DWORD bytesWritten = 0;
+            THROW_IF_WIN32_BOOL_FALSE(WriteFile(file.get(), buffer.data(), fileSize, &bytesWritten, nullptr));
+            THROW_WIN32_IF_MSG(ERROR_WRITE_FAULT, bytesWritten != fileSize, "failed to write");
+            buffer.clear();
         }
 
-        const auto& runs = row.Attributes().runs();
-        auto it = runs.begin();
-        const auto end = runs.end();
-        const auto last = end - 1;
-        til::CoordType oldX = 0;
-
-        for (; it != end; ++it)
+        if (!moreRowsRemaining)
         {
-            const auto attr = it->value.GetCharacterAttributes();
-            const auto hyperlinkId = it->value.GetHyperlinkId();
-            const auto fg = it->value.GetForeground();
-            const auto bg = it->value.GetBackground();
-            const auto ul = it->value.GetUnderlineColor();
+            break;
+        }
+    }
+}
 
-            if (previousAttr != attr)
+// Routine Description:
+// - Serializes one row the text buffer including ANSI escape code control sequences.
+// Arguments:
+// - row - A reference to the row being serialized.
+// - startX - The first column (inclusive) to include in the serialized content.
+// - endX - The last column (exclusive) to include in the serialized content.
+// - addLineBreak - Whether to add a line break at the end of the serialized row.
+// - buffer - A string to write the serialized row into.
+// - previousTextAttr - Used for tracking state across multiple calls to `SerializeRow` for sequential rows. The value will be mutated by the call. The initial call should contain `nullopt`, and subsequent calls should pass the value that was written by the previous call.

```

Comment:
Is there any particular way to notate out params (like `buffer`) and in/out params (like `previousTextAttr`)? 

Also for `previousTextAttr`, another option is to have it take a copy of the value and return a `TextAttribute` indicating the final state, which callers would then need to pass to subsequent calls; is that better? (maybe less spooky action at a distance? Mutating input params gives me the heebie jeebies 👻)